File: backend/entity/room.go

package entity

import (
    "gorm.io/gorm"
    "time"
)

type Room struct {
    gorm.Model
    RoomNumber  string    `gorm:"uniqueIndex"`
    Status      string  
    Image       string   
    BookingTime time.Time 

    // Foreign Keys
    RoomTypeID uint  
    StudentID  *uint 
    AdminID    uint  

    // Relations
    RoomType  *RoomType   `gorm:"foreignKey:RoomTypeID"`
    Student   *Student    `gorm:"foreignKey:StudentID"`
    Admin     *Admin      `gorm:"foreignKey:AdminID"`
    RoomAsset []RoomAsset `gorm:"foreignKey:RoomNumber;references:RoomNumber"`
}

File: frontend/src/interfaces/Room.ts
import type { RoomType } from "./RoomType";
import type { StudentInterface } from "./Student";
import type { AdminInterface } from "./Admin";
import type { RoomAsset } from "./RoomAsset";
import type { MeterInterface } from "./Meter";
import type { BillInterface } from "./Bill";

export interface RoomInterface {
  CreatedAt?: string;
  UpdatedAt?: string;
  DeletedAt?: string | null;

  ID?: number;
  RoomNumber: string;
  Status: string;
  Image?: string;
  BookingTime: string; // ISO date string

  // Foreign Keys
  RoomTypeID: number;
  StudentID: number | null;
  AdminID: number;

  // Relations
  RoomType?: RoomType;
  Student?: StudentInterface;
  Admin?: AdminInterface;
  RoomAsset?: RoomAsset[];
}

export interface RoomMeterInterface {
  ID: number;
  RoomNumber?: string;

  Student?: StudentInterface;       // ✅ array ของ student
  MeterRecords?: MeterInterface[];     // ✅ array ของ meter record
  Bills?: BillInterface[]; 
  
}
